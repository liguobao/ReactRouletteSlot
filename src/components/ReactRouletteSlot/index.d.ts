/*
 * @Author: wzi
 * @Date: 2018-10-01 10:47:26
 * @Last Modified by: wzi
 * @Last Modified time: 2018-10-01 11:31:20
 */
declare namespace ReactRouletteSlot {
    interface RouletteSlotDataItem {
        // 名称
        label: string | JSX.Element;
        // 图片
        img: string;
        // 商品 ID
        id: number;
    }
    type RouletteSlotData = RouletteSlotDataItem[];
}
