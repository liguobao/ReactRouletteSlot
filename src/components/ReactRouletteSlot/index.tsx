/*
 * @Author: wzi
 * @Date: 2018-10-01 10:23:08
 * @Last Modified by: wzi
 * @Last Modified time: 2019-01-07 17:03:05
 */
/// <reference path='index.d.ts' />
import React from 'react';
import {
    Wrapper,
    Item,
    LuckyButton,
    Line,
    Content,
    LuckyLabel,
    ContentItem,
    Img,
} from './style';
import { shouldUpdate } from '@common/decorator/decorator';
import { compose } from '@common/helper/compose';
import emptyEnhancer from '@common/HOC/Empty';
import errorBoundaryEnhancer from '@common/HOC/ErrorBoundary';
import Border from './Border';
const MIN_ROUND = 2;

@shouldUpdate({
    props: ['data', 'action'],
    state: ['pointer', 'run'],
})
export class ReactRouletteSlot extends React.Component<
    ReactRouletteSlotProps,
    ReactRouletteSlotState
> {
    state = {
        board: [[]],
        boardData: [],
        pointer: -1,
        target: null,
        round: 0,
        endRound: 0,
        luckyButtonPosition: {
            x: 0,
            y: 0,
        },
        run: false,
        row: null,
        col: null,
    };
    timer = null;
    componentWillMount() {
        this.dataHandler();
    }
    componentWillUnmount() {
        if (this.timer) {
            clearInterval(this.timer);
            this.timer = null;
        }
    }
    isActive = (position: number) => this.state.pointer === position;
    // 活动棋盘的数据
    getBoardData = () => this.state.boardData;

    // 寻找目标节点
    findTargetPointer = (target: number): DataItem =>
        this.getBoardData().find((cur) => {
            if (cur.type === 'button') {
                return false;
            }
            return cur.data.id === target;
        });
    // 获得几列
    getRow = (data: any[]) => (this.props.row || data.length < 9 ? 3 : 4);
    // 获得几行
    getCol = (row: number) => (this.props.data.length - 2 * row) / 2 + 2;
    // 获得按钮的坐标数据
    getLuckyButtonPosition = (data: RouletteSlotData, row: number) => {
        const x = row - 2;
        const y = (data.length - row * 2) / 2;
        this.setState({
            luckyButtonPosition: {
                x,
                y,
            },
        });
    };
    // 生成棋盘
    getBoard = (
        row: number,
        col: number,
        length: number,
        initData: RouletteSlotData
    ): [DataItem[][], DataItem[]] => {
        const board = [];
        const boardData = [];
        let position = 0;
        let Y = 0;
        for (let x = 0; x < col; x++) {
            board[x] = [];

            for (let y = 0; y < row; y++) {
                Y = y;
                if (x === 0) {
                    position = y;
                } else if (x === col - 1) {
                    position = row + col - 2 + y;
                    Y = row - y - 1;
                } else if (y === 0 || y === 1) {
                    if (y === 0) {
                        position = length - x;
                    } else {
                        position = row + x - 1;
                    }
                }
                if (position === -1) {
                    break;
                }

                board[x][Y] = {
                    position,
                    data: initData[position],
                    type: 'item',
                };
                position = -1;

                boardData.push(board[x][Y]);
            }
        }
        return [board, boardData];
    };
    // 加入抽奖按钮
    joinButton = (board: DataItem[][], boardData: DataItem[], row: number) => {
        const temp = board[1][1];
        const button: DataItem = { type: 'button' };
        board[1][1] = button;
        board[1][2] = temp;
        boardData.splice(row + 1, 0, button);
    };
    // 根据数据生成行数, 列数, 棋盘, 抽奖按钮的数据
    dataHandler = () => {
        const initData = this.props.data;
        const row = this.getRow(initData);
        const col = this.getCol(row);
        const [board, boardData] = this.getBoard(
            row,
            col,
            initData.length,
            initData
        );
        this.joinButton(board, boardData, row);
        this.setState({
            board,
            row,
            col,
            boardData,
        });

        this.getLuckyButtonPosition(initData, row);
    };
    // 请求开奖结果的回调
    onResultReturn = ({ data }: { data: number | string }) => {
        this.setState({
            target: data,
        });
    };
    // 请求数据
    onFetch() {
        setTimeout(() => {
            this.props.action(this.onResultReturn);
        }, 2000);
    }
    // 抽奖
    onClick = () => {
        if (!this.state.run) {
            // 点击启动跑马灯
            this.setState({
                pointer: 0,
                round: 0,
                endRound: 0,
                run: true,
            });
            this.onFetch();
            this.run();
        }
    };
    onSuccess = (target: DataItem) => {
        window.Alert.success(`恭喜您获得${target.data.label}`);
        return this.reset();
    };
    onFail = () => {
        window.Alert.success('请稍后重试');
        return this.reset();
    };
    getTime = (time: number, interval: number = -50) => {
        if (time < 100) {
            return 100;
        } else {
            return time + interval;
        }
    };
    position = () => {
        let pointer = this.state.pointer + 1;
        let round = this.state.round;
        if (pointer >= this.props.data.length) {
            pointer = 0;
            round++;
        }

        this.setState({
            pointer,
            round,
        });
        return [pointer, round];
    };
    reset = () => {
        this.setState({
            run: false,
        });
        return clearInterval(this.timer);
    };
    runBeforeEnd = (pointer: number, time: number) => {
        const target = this.findTargetPointer(this.state.target);
        if (this.state.endRound < 2) {
            if (pointer === target.position) {
                this.setState({ endRound: this.state.endRound + 1 });
            }
            return this.run(this.getTime(time));
        }

        if (!target) {
            return this.onFail();
        }
        if (target.position === pointer) {
            return this.onSuccess(target);
        }
        // 进入结束流程
        this.run(this.getTime(time, 30));
    };
    run = (time: number = 500) => {
        this.timer = setTimeout(() => {
            const [pointer, round] = this.position();
            if (round > MIN_ROUND && this.state.target !== null) {
                this.runBeforeEnd(pointer, time);
            } else {
                this.run(this.getTime(time));
            }
        }, time);
    };
    // 抽奖按钮
    LuckyButton = () => (
        <LuckyButton
            x={this.state.luckyButtonPosition.x}
            y={this.state.luckyButtonPosition.y}
            onClick={this.onClick}
            disable={this.state.run}
        >
            <LuckyLabel>
                <div>点击</div>
                <div>抽奖</div>
            </LuckyLabel>
        </LuckyButton>
    );

    // 物品节点
    Item = ({
        data,
        position,
    }: {
        position: number;
        data: RouletteSlotDataItem;
    }) => (
        <Item>
            <Content active={this.isActive(position)}>
                <ContentItem>
                    <Img src={data.img} />
                </ContentItem>
                <ContentItem>
                    <div>{data.label}</div>
                </ContentItem>
            </Content>
        </Item>
    );
    // 每个节点
    Grid = (item: DataItem, index: number) =>
        item.type === 'item' ? (
            <this.Item data={item.data} position={item.position} key={index} />
        ) : (
            <this.LuckyButton key={index} />
        );
    // 每一行
    Line = (line: DataItem[], key: number) => (
        <Line key={key}>{line.map(this.Grid)}</Line>
    );
    // 棋盘
    Board = () => (
        <Wrapper size={this.props.size}>
            <tbody>{this.state.board.map(this.Line)}</tbody>
        </Wrapper>
    );

    render() {
        return (
            <Border
                size={this.props.size}
                row={this.state.row}
                col={this.state.col}
                isRun={this.state.run}
            >
                <this.Board />
            </Border>
        );
    }
}
export default compose(
    emptyEnhancer(
        ({ data }: { data: RouletteSlotData }) =>
            !data || data.length === 0 || data.length % 2 !== 0,
        () => {
            window.Alert && window.Alert.fail('数据不符合要求');
            return null;
        }
    ),
    errorBoundaryEnhancer('角子机载入失败')
)(ReactRouletteSlot);
