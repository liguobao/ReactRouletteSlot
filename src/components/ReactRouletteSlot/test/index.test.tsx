/*
 * @Author: wzi
 * @Date: 2018-01-31 16:40:02
 * @Last Modified by: wzi
 * @Last Modified time: 2018-04-27 15:32:26
 */
import React from 'react';
import { ReactRouletteSlot } from '@components/ReactRouletteSlot';
import '@common/helper/Alert';
import { mount } from 'enzyme';
import 'jest-styled-components';
import toJson from 'enzyme-to-json';
const data = [
    { id: 1000, img: 'http://dummyimage.com/30x30', label: 'Larry' },
    { id: 1001, img: 'http://dummyimage.com/30x30', label: 'Joseph' },
    { id: 1003, img: 'http://dummyimage.com/30x30', label: 'Paul' },
    { id: 1004, img: 'http://dummyimage.com/30x30', label: 'Ronald' },
    { id: 1005, img: 'http://dummyimage.com/30x30', label: 'Helen' },
    { id: 1006, img: 'http://dummyimage.com/30x30', label: 'Maria' },
    { id: 1007, img: 'http://dummyimage.com/30x30', label: 'Mark' },
    { id: 1008, img: 'http://dummyimage.com/30x30', label: 'Mark' },
    { id: 1009, img: 'http://dummyimage.com/30x30', label: 'Carol' },
    { id: 1010, img: 'http://dummyimage.com/30x30', label: 'Ronald' },
    { id: 1011, img: 'http://dummyimage.com/30x30', label: 'Nancy' },
    { id: 1012, img: 'http://dummyimage.com/30x30', label: 'Michelle' },
];
jest.useFakeTimers();

describe('Loading', () => {
    let _wrapper;
    beforeAll(() => {
        const action = (cb) => {
            cb({ data: 1011 });
        };

        _wrapper = mount(<ReactRouletteSlot data={data} action={action} />);
    });
    test('整体', () => {
        expect(toJson(_wrapper)).toMatchSnapshot();
    });
    test('点击抽奖', () => {
        const instance = _wrapper.instance();
        instance.onClick();
        jest.advanceTimersByTime(20000);
    });
});
