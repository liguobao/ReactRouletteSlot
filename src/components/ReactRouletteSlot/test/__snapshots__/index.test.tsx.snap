// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading 整体 1`] = `
.c9 {
  width: 100%;
  height: 100%;
  text-align: center;
  table-layout: fixed;
  background: #ae184e;
  position: absolute;
  top: 20px;
  left: 20px;
}

.c10 {
  padding: 2px;
}

.c12 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c12:first-child {
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

.c11 {
  background: #fdf29a;
  height: 100%;
  width: 100%;
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c14 {
  opacity: 1;
  padding: 2px;
}

.c14:hover {
  cursor: pointer;
}

.c15 {
  width: 100%;
  height: 100%;
  background: #f64222;
  color: white;
  font-size: 40px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c13 {
  margin-top: 20%;
}

.c0 {
  border-radius: 10px;
  border-collapse: separate;
  position: relative;
  width: calc(100% + 40px);
  height: calc(100% + 40px);
  margin: 0 auto;
  background: #fc5a38;
}

.c1 {
  width: 10px;
  height: 10px;
  border-radius: 50%;
  position: absolute;
  top: 5px;
  left: NaNpx;
  bottom: px;
  right: px;
  background: #4999da;
}

.c2 {
  width: 10px;
  height: 10px;
  border-radius: 50%;
  position: absolute;
  top: 5px;
  left: NaNpx;
  bottom: px;
  right: px;
  background: yellow;
}

.c3 {
  width: 10px;
  height: 10px;
  border-radius: 50%;
  position: absolute;
  top: px;
  left: NaNpx;
  bottom: 5px;
  right: px;
  background: #4999da;
}

.c4 {
  width: 10px;
  height: 10px;
  border-radius: 50%;
  position: absolute;
  top: px;
  left: NaNpx;
  bottom: 5px;
  right: px;
  background: yellow;
}

.c5 {
  width: 10px;
  height: 10px;
  border-radius: 50%;
  position: absolute;
  top: NaNpx;
  left: 5px;
  bottom: px;
  right: px;
  background: #4999da;
}

.c6 {
  width: 10px;
  height: 10px;
  border-radius: 50%;
  position: absolute;
  top: NaNpx;
  left: 5px;
  bottom: px;
  right: px;
  background: yellow;
}

.c7 {
  width: 10px;
  height: 10px;
  border-radius: 50%;
  position: absolute;
  top: NaNpx;
  left: px;
  bottom: px;
  right: 5px;
  background: #4999da;
}

.c8 {
  width: 10px;
  height: 10px;
  border-radius: 50%;
  position: absolute;
  top: NaNpx;
  left: px;
  bottom: px;
  right: 5px;
  background: yellow;
}

<ReactRouletteSlot
  action={[Function]}
  data={
    Array [
      Object {
        "id": 1000,
        "img": "http://dummyimage.com/30x30",
        "label": "Larry",
      },
      Object {
        "id": 1001,
        "img": "http://dummyimage.com/30x30",
        "label": "Joseph",
      },
      Object {
        "id": 1003,
        "img": "http://dummyimage.com/30x30",
        "label": "Paul",
      },
      Object {
        "id": 1004,
        "img": "http://dummyimage.com/30x30",
        "label": "Ronald",
      },
      Object {
        "id": 1005,
        "img": "http://dummyimage.com/30x30",
        "label": "Helen",
      },
      Object {
        "id": 1006,
        "img": "http://dummyimage.com/30x30",
        "label": "Maria",
      },
      Object {
        "id": 1007,
        "img": "http://dummyimage.com/30x30",
        "label": "Mark",
      },
      Object {
        "id": 1008,
        "img": "http://dummyimage.com/30x30",
        "label": "Mark",
      },
      Object {
        "id": 1009,
        "img": "http://dummyimage.com/30x30",
        "label": "Carol",
      },
      Object {
        "id": 1010,
        "img": "http://dummyimage.com/30x30",
        "label": "Ronald",
      },
      Object {
        "id": 1011,
        "img": "http://dummyimage.com/30x30",
        "label": "Nancy",
      },
      Object {
        "id": 1012,
        "img": "http://dummyimage.com/30x30",
        "label": "Michelle",
      },
    ]
  }
>
  <Board
    col={4}
    isRun={false}
    row={4}
  >
    <styled.div>
      <div
        className="c0"
      >
        <Component>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="0"
            left={NaN}
            right={null}
            top={5}
          >
            <div
              className="c1"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="1"
            left={NaN}
            right={null}
            top={5}
          >
            <div
              className="c2"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="2"
            left={NaN}
            right={null}
            top={5}
          >
            <div
              className="c1"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="3"
            left={NaN}
            right={null}
            top={5}
          >
            <div
              className="c2"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="4"
            left={NaN}
            right={null}
            top={5}
          >
            <div
              className="c1"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="5"
            left={NaN}
            right={null}
            top={5}
          >
            <div
              className="c2"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="6"
            left={NaN}
            right={null}
            top={5}
          >
            <div
              className="c1"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="7"
            left={NaN}
            right={null}
            top={5}
          >
            <div
              className="c2"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="8"
            left={NaN}
            right={null}
            top={5}
          >
            <div
              className="c1"
            />
          </styled.div>
          <styled.div
            bottom={5}
            isOdd={true}
            isRun={false}
            key="0"
            left={NaN}
            right={null}
            top={null}
          >
            <div
              className="c3"
            />
          </styled.div>
          <styled.div
            bottom={5}
            isOdd={false}
            isRun={false}
            key="1"
            left={NaN}
            right={null}
            top={null}
          >
            <div
              className="c4"
            />
          </styled.div>
          <styled.div
            bottom={5}
            isOdd={true}
            isRun={false}
            key="2"
            left={NaN}
            right={null}
            top={null}
          >
            <div
              className="c3"
            />
          </styled.div>
          <styled.div
            bottom={5}
            isOdd={false}
            isRun={false}
            key="3"
            left={NaN}
            right={null}
            top={null}
          >
            <div
              className="c4"
            />
          </styled.div>
          <styled.div
            bottom={5}
            isOdd={true}
            isRun={false}
            key="4"
            left={NaN}
            right={null}
            top={null}
          >
            <div
              className="c3"
            />
          </styled.div>
          <styled.div
            bottom={5}
            isOdd={false}
            isRun={false}
            key="5"
            left={NaN}
            right={null}
            top={null}
          >
            <div
              className="c4"
            />
          </styled.div>
          <styled.div
            bottom={5}
            isOdd={true}
            isRun={false}
            key="6"
            left={NaN}
            right={null}
            top={null}
          >
            <div
              className="c3"
            />
          </styled.div>
          <styled.div
            bottom={5}
            isOdd={false}
            isRun={false}
            key="7"
            left={NaN}
            right={null}
            top={null}
          >
            <div
              className="c4"
            />
          </styled.div>
          <styled.div
            bottom={5}
            isOdd={true}
            isRun={false}
            key="8"
            left={NaN}
            right={null}
            top={null}
          >
            <div
              className="c3"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="0"
            left={5}
            right={null}
            top={NaN}
          >
            <div
              className="c5"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="1"
            left={5}
            right={null}
            top={NaN}
          >
            <div
              className="c6"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="2"
            left={5}
            right={null}
            top={NaN}
          >
            <div
              className="c5"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="3"
            left={5}
            right={null}
            top={NaN}
          >
            <div
              className="c6"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="4"
            left={5}
            right={null}
            top={NaN}
          >
            <div
              className="c5"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="5"
            left={5}
            right={null}
            top={NaN}
          >
            <div
              className="c6"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="6"
            left={5}
            right={null}
            top={NaN}
          >
            <div
              className="c5"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="7"
            left={5}
            right={null}
            top={NaN}
          >
            <div
              className="c6"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="8"
            left={5}
            right={null}
            top={NaN}
          >
            <div
              className="c5"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="0"
            left={null}
            right={5}
            top={NaN}
          >
            <div
              className="c7"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="1"
            left={null}
            right={5}
            top={NaN}
          >
            <div
              className="c8"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="2"
            left={null}
            right={5}
            top={NaN}
          >
            <div
              className="c7"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="3"
            left={null}
            right={5}
            top={NaN}
          >
            <div
              className="c8"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="4"
            left={null}
            right={5}
            top={NaN}
          >
            <div
              className="c7"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="5"
            left={null}
            right={5}
            top={NaN}
          >
            <div
              className="c8"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="6"
            left={null}
            right={5}
            top={NaN}
          >
            <div
              className="c7"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={false}
            isRun={false}
            key="7"
            left={null}
            right={5}
            top={NaN}
          >
            <div
              className="c8"
            />
          </styled.div>
          <styled.div
            bottom={null}
            isOdd={true}
            isRun={false}
            key="8"
            left={null}
            right={5}
            top={NaN}
          >
            <div
              className="c7"
            />
          </styled.div>
        </Component>
        <Component>
          <styled.table>
            <table
              className="c9"
            >
              <tbody>
                <styled.tr
                  key="0"
                >
                  <tr
                    className=""
                  >
                    <Component
                      data={
                        Object {
                          "id": 1000,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Larry",
                        }
                      }
                      key="0"
                      position={0}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Larry
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                    <Component
                      data={
                        Object {
                          "id": 1001,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Joseph",
                        }
                      }
                      key="1"
                      position={1}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Joseph
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                    <Component
                      data={
                        Object {
                          "id": 1003,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Paul",
                        }
                      }
                      key="2"
                      position={2}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Paul
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                    <Component
                      data={
                        Object {
                          "id": 1004,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Ronald",
                        }
                      }
                      key="3"
                      position={3}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Ronald
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                  </tr>
                </styled.tr>
                <styled.tr
                  key="1"
                >
                  <tr
                    className=""
                  >
                    <Component
                      data={
                        Object {
                          "id": 1012,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Michelle",
                        }
                      }
                      key="0"
                      position={11}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Michelle
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                    <Component
                      key="1"
                    >
                      <styled.td
                        disable={false}
                        onClick={[Function]}
                        x={2}
                        y={2}
                      >
                        <td
                          className="c14"
                          colSpan={2}
                          onClick={[Function]}
                          rowSpan={2}
                          x={2}
                          y={2}
                        >
                          <styled.div>
                            <div
                              className="c15"
                            >
                              点击抽奖
                            </div>
                          </styled.div>
                        </td>
                      </styled.td>
                    </Component>
                    <Component
                      data={
                        Object {
                          "id": 1005,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Helen",
                        }
                      }
                      key="2"
                      position={4}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Helen
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                  </tr>
                </styled.tr>
                <styled.tr
                  key="2"
                >
                  <tr
                    className=""
                  >
                    <Component
                      data={
                        Object {
                          "id": 1011,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Nancy",
                        }
                      }
                      key="0"
                      position={10}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Nancy
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                    <Component
                      data={
                        Object {
                          "id": 1006,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Maria",
                        }
                      }
                      key="1"
                      position={5}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Maria
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                  </tr>
                </styled.tr>
                <styled.tr
                  key="3"
                >
                  <tr
                    className=""
                  >
                    <Component
                      data={
                        Object {
                          "id": 1010,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Ronald",
                        }
                      }
                      key="0"
                      position={9}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Ronald
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                    <Component
                      data={
                        Object {
                          "id": 1009,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Carol",
                        }
                      }
                      key="1"
                      position={8}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Carol
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                    <Component
                      data={
                        Object {
                          "id": 1008,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Mark",
                        }
                      }
                      key="2"
                      position={7}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Mark
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                    <Component
                      data={
                        Object {
                          "id": 1007,
                          "img": "http://dummyimage.com/30x30",
                          "label": "Mark",
                        }
                      }
                      key="3"
                      position={6}
                    >
                      <styled.td>
                        <td
                          className="c10"
                        >
                          <styled.span
                            active={false}
                          >
                            <span
                              className="c11"
                            >
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <styled.img
                                    src="http://dummyimage.com/30x30"
                                  >
                                    <img
                                      className="c13"
                                      src="http://dummyimage.com/30x30"
                                    />
                                  </styled.img>
                                </span>
                              </styled.span>
                              <styled.span>
                                <span
                                  className="c12"
                                >
                                  <div>
                                    Mark
                                  </div>
                                </span>
                              </styled.span>
                            </span>
                          </styled.span>
                        </td>
                      </styled.td>
                    </Component>
                  </tr>
                </styled.tr>
              </tbody>
            </table>
          </styled.table>
        </Component>
      </div>
    </styled.div>
  </Board>
</ReactRouletteSlot>
`;
